cccatalog-api/cccatalog/api/utils/watermark.py

```diff
@@ -22,63 +176,53 @@ def _open_image(url):
         print('Error loading image data')
 
 
-def _create_frame_for_image(image):
+def _print_attribution_on_image(img, image_info):
     """
-    Creates an image as a frame from another image.
+    Add a frame around the image and put the attribution text on the bottom
+    :param img: the image to frame and attribute
+    :param image_info: the information about a particular image
+    :return: return the framed and attributed image
     """
-    size = image.size
-    width = size[0]
-    height = size[1]
-    img = Image.new(
-        "RGB",
-        (width + horizontal_margin, height + vertical_margin),
-        frame_color
-    )
-    return img
 
+    width, height = img.size
+    smaller_dimension = _smaller_dimension(width, height)
 
-def _place_image_inside_frame(image):
-    frame = _create_frame_for_image(image)
-    copy = image.copy()
-    top_margin = int(vertical_margin / 4)
-    left_margin = int(horizontal_margin / 2)
-    frame.paste(copy, (left_margin, top_margin))
+    if smaller_dimension is Dimension.NONE:
+        margin = round(MARGIN_RATIO * min(width, height))
+        font_size = round(FONT_RATIO * min(width, height))
+        new_width = width
+    else:
+        margin = round(MARGIN_RATIO * BREAKPOINT_DIMENSION)
+        font_size = round(FONT_RATIO * BREAKPOINT_DIMENSION)
+        new_width = BREAKPOINT_DIMENSION \
+            if Dimension.WIDTH in smaller_dimension \
+            else width
 
-    return frame
+    font = ImageFont.truetype(_get_font_path(), size=font_size)
 
+    text = _get_attribution_text(image_info)
+    text = _fit_in_width(text, font, new_width)
+    _, attribution_height = font.getsize_multiline(text)
 
-def _full_license(image_info):
-    _license = image_info['license'].upper()
-    license_version = image_info['license_version'].upper()
-    license_text = "{0} {1}".format(_license, license_version)
-    return license_text if _license == "cc0" else "CC {0}".format(license_text)
-
+    frame_width = margin + new_width + margin
+    frame_height = margin + height + margin + attribution_height + margin
+    left_margin = (frame_width - width) // 2
 
-def _print_attribution_for_image_on_frame(image_info, image, frame):
-    vertical_margin_to_image = 16  # vertical margin between image and text
+    frame = _create_frame((frame_width, frame_height,))
+    _frame_image(img, frame, left_margin, margin)
 
-    try:
-        font = ImageFont.truetype('SourceSansPro-Bold.ttf', size=18)
-    except OSError:
```

Comment:
Please restore this try/except block and remove _get_font_path. The fonts vary by deployment environment (as at this time we do not deploy the frontend through Docker). We can't hardcode any fonts or else it will crash if unavailable.

Reply:
I'll do that momentarily. Can you help me understand how the deployment of the frontend with/without Docker affects how the API watermarks the images?

Reply:
If you see the `Dockerfile`, it contains the line
```
ADD cccatalog/api/utils/fonts/SourceSansPro-Bold.ttf /usr/share/fonts/truetype/SourceSansPro-Bold.ttf
```
which means that for Pillow to load the font, we had to **install** it onto the Docker system. The function `_get_font_path()` allows us to avoid doing that anytime we want to switch over to the more accurate monospace font-set, or any font that we may want to switch to in the future.

Also worth noting is that `ImageFont.truetype()` returns a `FreeTypeFont` whereas `ImageFont.load_default()` or `None` default to a simple `ImageFont`.
![image](https://user-images.githubusercontent.com/16580576/54556221-6a824380-49de-11e9-9a0e-d402616ead97.png)
As you can see `ImageFont` does not have a `getsize_multiline()` function that is essential to the frame dimension calculation.

Can I suggest a compromise that we keep the `_get_font_path()` function, only to be used in the event of the `OSError` exception, just to allow the rest of the flow to complete normally? As your previous comment said, this branch will only be executed in CI. Never else.

@aldenstpage I'll update the PR with your inputs.